autopilot/autopilot.go

```diff
@@ -131,29 +138,57 @@ func (ap *Autopilot) Run() error {
 		ap.c.applyConsensusState(cfg, cs)
 
 		// perform maintenance
-		_ = ap.c.performContractMaintenance(cfg) // TODO: handle error
+		err = ap.c.performContractMaintenance(cfg)
+		if err != nil {
+			ap.logger.Errorf("contract maintenance failed, err: %v", err)
+			continue
+		}
 
 		// migration
-		_ = ap.m.UpdateContracts() // TODO: handle error
+		err = ap.m.UpdateContracts()
+		if err != nil {
+			ap.logger.Errorf("update contracts failed, err: %v", err)
+		}
 		ap.m.TryPerformMigrations()
 	}
 }
 
-func (ap *Autopilot) Stop() {
+func (ap *Autopilot) Stop() error {
 	ap.ticker.Stop()
 	close(ap.stopChan)
+
+	errs := make([]error, len(ap.onStopFns))
+	for i, onStopFn := range ap.onStopFns {
+		errs[i] = onStopFn()
+	}
+	return joinErrors(errs)
 }
 
 // New initializes an Autopilot.
-func New(store Store, bus Bus, worker Worker, heartbeat time.Duration) (*Autopilot, error) {
+func New(store Store, bus Bus, worker Worker, dir string, heartbeat time.Duration) (*Autopilot, error) {
+	// instantiate the logger
+	logger, closeFn, err := newLogger(filepath.Join(dir, logFileName))
+	if err != nil {
+		return nil, err
+	}
+
+	onStopFn := func() (err error) {
+		err = logger.Sync()
+		closeFn()
+		return
+	}
```

Comment:
this is a siad/skyd pattern that I am trying to avoid replicating. 😅

The rule of thumb is: *Don't call constructors within a constructor*. Doing so is convenient for the caller (since they can pass one `dir` argument instead of a bunch of interfaces), but it has three big downsides:
- It prevents you from using interfaces. In this case, we are stuck with a logger that writes to a file, even during testing.
- It forces the constructed object to be responsible for cleaning up its dependencies. (Should an autopilot *really* be responsible for closing a logger?)
- It prevents you from sharing a dependency. What if we want all of our subsystems to share a logger?

There are exceptions, of course. But to me, this is a pretty clear example of where we should be passing a `Logger` interface instead of calling a constructor within a constructor.

I think the constructor-within-a-constructor pattern derives from a more general design philosophy of building systems out of "actors," by which I mean "black boxes that have a lot of state and make a lot of decisions." I'm moving away from that philosophy, and towards systems built from "dumb" components, which have less state and make fewer decisions. Essentially, the goal is to *centralize the complexity* of the system by moving it out of the individual components.

Reply:
That makes sense 👍 - couldn't pass an interface because the guys at Uber decided to [remove the interface](https://github.com/uber-go/zap/issues/381) and make it a struct. Should be fine though, it actually cleaned up the cleanup function, thanks!